feat(linter/plugins): `oxlint` export types

overlookmotel opened the pull request:
`oxlint` package export TS types. `defineRule` and `definePlugin` now give you type-safety and intellisense.

Writing plugins in TS is only supported on versions of NodeJS which support type-stripping. Have not added a library (e.g. Jiti) to support older NodeJS versions.

Probably some of the types are wrong. But I think this is a decent start.

"Are the types wrong" (added in #14177) is reporting an error in `index.d.ts`:

```ts
import { VisitorObject as Visitor } from "../dist/generated/visit/visitor.js";
//                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

because the file in `dist/generated/visit` is actually called `visitor.d.ts` (`.d.ts`, not `.js`).

I'm not sure if that's a mistake in Rolldown's bundling, "Are the types wrong" being too strict, or a genuine problem.

But regardless, TS seems to pick up the types from `visitor.d.ts`, so I think we can probably merge this and look into / fix it later.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14163?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#14163** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14163?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14163?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#14162** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14162?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#14159** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/14159?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`
<hr />

## How to use the Graphite Merge Queue

Add either label to this PR to merge it via the merge queue:
- <kbd>0-merge</kbd> - adds this PR to the back of the merge queue
- <kbd>hotfix</kbd> - for urgent hot fixes, skip the queue and merge this PR next

You must have a Graphite account in order to use the merge queue. Sign up using [this link](https://app.graphite.dev/invite/github/oxc-project?ref=merge-queue-instructions-in-stack-comment&prId=14163).

<h2></h2>
<sub>

An organization admin has enabled the [Graphite Merge Queue](https://graphite.dev/docs/graphite-merge-queue) in this repository.

</sub>

<sub>

Please do not merge from GitHub as this will restart CI on PRs being processed by the merge queue.

</sub>

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

overlookmotel commented on apps/oxlint/package.json:
```diff
@@ -34,9 +35,13 @@
   "devDependencies": {
     "eslint": "^9.36.0",
     "execa": "^9.6.0",
+    "jiti": "^2.6.0",
```
`jiti` is required in tests for ESLint compatibility. ESLint requires `jiti` to be present as a dependency to load plugins from `.ts` files. So `devDependencies` is the right place for it.

overlookmotel commented on apps/oxlint/src-js/index.ts:
```diff
@@ -2,6 +2,8 @@ import type { Context } from './plugins/context.ts';
 import type { CreateOnceRule, Plugin, Rule } from './plugins/load.ts';
 import type { BeforeHook, Visitor, VisitorWithHooks } from './plugins/types.ts';
 
+export type { Context, Plugin, Rule, Visitor, VisitorWithHooks };
```
Yes, I agree, I think need to export more types. I'll make that change now.

overlookmotel replied:
Done. All types are exported now. I've checked the generated `index.d.ts` file and none are missing.

overlookmotel commented:
I've changed this so `oxlint` package now exports all types used in the external interface. I've checked carefully and the exports don't miss any.

#14200 converts all the plugins in text fixtures to TS, and the only oddity it threw up is this one:

https://github.com/oxc-project/oxc/pull/14200#pullrequestreview-3276126224

If that's OK, I think we're good to go.

camc314 reviewed (approved):
do we need to update the `npm/oxlint` package.json with the exports?

camc314 commented:
> @camc314 Would very much appreciate your help to improve the types / advise on whether more types should be exported, when you have time.
> 
> Please can you build `dist` folder locally, and check that the `index.d.ts` file looks OK?

yes looked and all looks ok, i'm not certain the types for `import * as ESTree from '@oxc-project/types';` are correct, but tricky to tell without a release

overlookmotel committed:
feat(linter/plugins): `oxlint` export types (#14163)

`oxlint` package export TS types. `defineRule` and `definePlugin` now give you type-safety and intellisense.

Writing plugins in TS is only supported on versions of NodeJS which support type-stripping. Have not added a library (e.g. Jiti) to support older NodeJS versions.

Probably some of the types are wrong. But I think this is a decent start.

"Are the types wrong" (added in #14177) is reporting an error in `index.d.ts`:

```ts
import { VisitorObject as Visitor } from "../dist/generated/visit/visitor.js";
//                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

because the file in `dist/generated/visit` is actually called `visitor.d.ts` (`.d.ts`, not `.js`).

I'm not sure if that's a mistake in Rolldown's bundling, "Are the types wrong" being too strict, or a genuine problem.

But regardless, TS seems to pick up the types from `visitor.d.ts`, so I think we can probably merge this and look into / fix it later.
